query-engine/connector-test-kit-rs/query-test-macros/src/connector_test.rs

```diff
@@ -0,0 +1,116 @@
+        return syn::Error::new_spanned(
+            test_function.sig,
+            "connector test functions must take exactly one argument: `runner: &Runner`.",
+        )
+        .to_compile_error()
+        .into();
+    }
+
+    if test_function.sig.asyncness.is_none() {
+        return syn::Error::new_spanned(test_function.sig, "connector test functions must be async")
+            .to_compile_error()
+            .into();
+    }
+
+    // The shell function retains the name of the original test definition.
+    let test_fn_ident = test_function.sig.ident.clone();
+
+    // Rename original test function to run_<orig_name>.
+    let runner_fn_ident = Ident::new(&format!("run_{}", test_fn_ident.to_string()), Span::call_site());
+    test_function.sig.ident = runner_fn_ident.clone();
+
+    // The test database name is the name used as the database for data source rendering.
+    // Combination of test name and test mod name.
+    let test_name = test_fn_ident.to_string();
+    let suite_name = args.suite.expect("A test must have a test suite.");
+    let test_database = format!("{}_{}", suite_name, test_name);
+    let capabilities: Vec<_> = args
+        .capabilities
+        .idents
+        .into_iter()
+        .map(|cap| {
+            quote! {
+                ConnectorCapability::#cap
+            }
+        })
+        .collect();
+
+    // The actual test is a shell function that gets the name of the original function,
+    // which is then calling `{orig_name}_run` in the end (see `runner_fn_ident`).
+    let test = quote! {
+        #[test]
+        fn #test_fn_ident() {
+            let config = &query_tests_setup::CONFIG;
+            let enabled_connectors = vec![
+                #connectors
+            ];
+
+            let capabilities: Vec<ConnectorCapability> = vec![
+                #(#capabilities),*
+            ];
+
+            if ConnectorTag::should_run(&config, &enabled_connectors, &capabilities, #test_name) {
+                let template = #handler();
+                let datamodel = query_tests_setup::render_test_datamodel(config, #test_database, template);
+
+                query_tests_setup::run_with_tokio(async move {
+                    tracing::debug!("Used datamodel:\n {}", datamodel.clone().yellow());
+                    let runner = Runner::load(config.runner(), datamodel.clone()).await.unwrap();
+                    query_tests_setup::setup_project(&datamodel).await.unwrap();
+                    #runner_fn_ident(&runner).await.unwrap();
+                }.with_subscriber(test_tracing_subscriber(std::env::var("LOG_LEVEL").unwrap_or("info".to_string()))));
+            }
+        }
+
+        #test_function
+    };
+
+    test.into()
+}
+
+fn quote_connector(tag: ConnectorTag) -> proc_macro2::TokenStream {
+    let (connector, version) = tag.as_parse_pair();
+
+    match version {
+        Some(version) => quote! {
+            ConnectorTag::try_from((#connector, Some(#version))).unwrap()
```

Comment:
This is fairly idiotic and build at the very start of my macro adventure. Can be redone later.